stdlib/public/core/Integers.swift.gyb

```diff
@@ -3279,210 +3250,436 @@ public struct DoubleWidth<T : FixedWidthInteger>
+      return (n < _storage.low.countRepresentedWords) ?
+        _storage.low._word(at: n) :
+        _storage.high._word(at: n - _storage.low.countRepresentedWords)
     }
-    // TODO: move to Int128 just like init(_builtinIntegerLiteral:) ?
-    let wordsInT = T.bitWidth / ${word_bits}
-    return (n < wordsInT) ?
-      _storage.low._word(at: n) :
-      _storage.high._word(at: n - wordsInT)
   }
 
   public static var isSigned: Bool {
-    return T.isSigned
+    return Base.isSigned
   }
 
   // fixed width
   //
-  public static var max: DoubleWidth<T> {
+  public static var max: DoubleWidth {
     return self.init((High.max, Low.max))
   }
 
-  public static var min: DoubleWidth<T> {
+  public static var min: DoubleWidth {
     return self.init((High.min, Low.min))
   }
 
-  public static var bitWidth : Int { return 2 * T.bitWidth }
+  public static var bitWidth: Int {
+    return 2 * Base.bitWidth
+  }
 
 % # This covers + and -
 % for x in binaryArithmetic['Numeric'][:2]:
-%   highAffectedByLowOverflow = 'T.max' if x.operator == '+' else 'T.min'
-  public func ${x.name}ReportingOverflow(_ rhs: DoubleWidth<T>)
-    -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
+%   highAffectedByLowOverflow = 'Base.max' if x.operator == '+' else 'Base.min'
+  public func ${x.name}ReportingOverflow(_ rhs: DoubleWidth)
+    -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
     let (low, lowOverflow) =
       _storage.low.${x.name}ReportingOverflow(rhs._storage.low)
     let (high, highOverflow) =
       _storage.high.${x.name}ReportingOverflow(rhs._storage.high)
     let isLowOverflow = lowOverflow == .overflow
-    let result = (high ${x.operator} (isLowOverflow ? 1 : 0), low)
+    let result = (high &${x.operator} (isLowOverflow ? 1 : 0), low)
     let overflow = ArithmeticOverflow(
       highOverflow == .overflow ||
       high == ${highAffectedByLowOverflow} && isLowOverflow
     )
-    return (partialValue: DoubleWidth<T>(result),
+    return (partialValue: DoubleWidth(result),
       overflow: overflow)
   }
 % end
 
-
   public func multipliedReportingOverflow(
-    by rhs: DoubleWidth<T>
-  ) -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
-    let isNegative = (self < DoubleWidth<T>()) != (rhs < DoubleWidth<T>())
-
-    func mul(_ x: Low, _ y: Low, _ carry: Low) -> (partial: Low, carry: Low) {
-      let pair = x.multipliedFullWidth(by: y)
-      let t = DoubleWidth<Low>(pair) + DoubleWidth<Low>((0, carry))
-      return (partial: t._storage.low, carry: t._storage.high)
-    }
+    by rhs: DoubleWidth
+  ) -> (partialValue: DoubleWidth, overflow: ArithmeticOverflow) {
+    let (carry, product) = multipliedFullWidth(by: rhs)
+    let result = DoubleWidth(extendingOrTruncating: product)
+    
+    let isNegative = (self < (0 as DoubleWidth)) != (rhs < (0 as DoubleWidth))
+    let didCarry = isNegative ? carry != ~0 : carry != 0
```

Comment:
Yikes, this is where the heterogeneous operators really start biting. Presumably, you didn't mean `~0` as in `Int(-1)`.

```swift
let didCarry = isNegative ? carry != ~(0 as DoubleWidth) : carry != (0 as DoubleWidth)
```